### DISCUSSION TITLE
Bitmap data from flash

### DISCUSSION THREAD

**Discussion by Marvin1DE** [RudolphRiedel/FT800-FT813#116] [Q&A] [answered]
Hey nochmal,

Also ich hab jetzt die letzten tage funktionen geschrieben die ich von BASCOM in C abgewandelt habe, müsste jetzt die richtigen datein im flash haben aber wie nutzte ich diese am effektivsten (Bilder und später fonts)?

Es fehlt zwar noch das ein oder andere wie die Liste an datein die man Lesen sollte und nicht nur eine aber soweit müsste jetzt die datei gelesen in den ram g geworfen werden, dann in den flash geschoben werden,

Hier die funktionen:

[main.txt](https://github.com/RudolphRiedel/FT800-FT813/files/14612383/main.txt)

Hier versuche ich das Bild zu nutzen:
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/9b503efb-1e58-4a3f-bfb6-61a30446127f)
bild:
Breite = 20
Laenge = 20
auf position:
x = 100
y = 100
Format:
ASTC_4X4 formatiert

**Comment by RudolphRiedel** [comment, marked as answer]
Moin,

das EVE_cmd_setbitmap() sieht verdächtig aus, bzw. die Adresse.
Seltsamerweise muss die Adresse nämlich in Blöcken von 32 Bytes angegeben werden wenn das Bild im Flash liegt.
Probier das mal so:
EVE_cmd_setbitmap_burst(0x800000 | (4096/32), EVE_ASTC_4x4, 20, 20);

**Reply by Marvin1DE** [reply]
Hab EVE_cmd_setbitmap_burst(0x800000 | (4096/32), EVE_ASTC_4x4, 20, 20); und
EVE_cmd_setbitmap_burst(0x801000 | (4096/32), EVE_ASTC_4x4, 20, 20); versucht.
Leider bisher kein erfolg muss schauen was genau ich falsch mache und probiere etwas rum.
Danke für deine Hilfe
MFG

edit: Ich habe auch daran gedacht es richtig in den Ram zu schreiben
EVE_memWrite_sram_buffer(0x001000 | (4096/32),transferbyte,400);
transferbyte ist ein [4096] array das 400 elementen des Bildes aufgefüllt würde und dann in den RAM_G geht durch diesen command
edit2: Das Bild ist auch im RAM_G, habe es getestet das bild wird angezeigt wenn ich es aus dem ram_g nehme
EVE_cmd_setbitmap_burst(0x001000 | (4096/32), EVE_ASTC_4x4, 20, 20);

**Reply by RudolphRiedel** [reply]
Ich meine das war jetzt nur geraten, welche Adresse steht denn konkret in der .map Datei für das Bild?
Und hast Du auch EVE_init_flash() benutzt um das Flash von Normal in den Fast Modus zu bringen?

**Reply by Marvin1DE** [reply]
hab erst meine eingene funktion geschrieben, hatte dann gesehen das diese bereits existiert und benutzte nun diese, init & fast flash geben keine fehler zurück.
im bezug auf .map datei, ich habe die datei als .bin abgespeichert und die konvertierte bild.raw umbenannt -> bild.bin, haben dies bei bascom getan gibt es einen effizienteren weg als es mit der .bin, ist die dateiendung vlt auch irrelevant? oder muss ich wenn ich das so mache irgendwas beachten?

**Reply by Marvin1DE** [reply]
in der .bin datei ist jedenfalls auch nur 25 mal das hier "FC FD FF FF FF FF FF FF 1B 1B FF FF 00 00 FF FF" ist auch nur ein grüner block zum test.
Hab EVE_cmd_flashupdate vergessen...... mein fehler, ist mir gerade aufgefallen als ich mehr informationen raus suchen wollte
my bad.

**Reply by Marvin1DE** [reply]
![grafik](https://github.com/RudolphRiedel/FT800-FT813/assets/118083130/cca4b4b4-6d2e-45cf-830a-a3b73f50fd53)
dennoch geht es nicht mach noch irgendetwas falsch

**Reply by Marvin1DE** [reply]
mit:
EVE_cmd_setbitmap_burst(0x800000 | (4096/32), EVE_ASTC_4x4, 20, 20); gehts jetzt doch jedoch hab ich das bild auf 0x801000 gespeichert bin etwas verwirrt wie es genau funktioniert mit dem abspeichern und anzeigen denn als ich auf 0x802000 gespeichert habe und EVE_cmd_setbitmap_burst(0x801000 | (4096/32),...) gesetzt habe ging es nicht mehr.

**Reply by RudolphRiedel** [reply]
Also der Name ist zunächst egal, es kann nur sein, dass EAB Bilder so erkennt und richtig im Flash platziert.
0x801000 ist die physikalische Adresse im Flash, also ab Flash-Basis + 4096, so ist die Adresse mit EVE_cmd_flashupdate() oder EVE_cmd_flashread() zu verwenden.
EVE_cmd_setbitmap() braucht die Adresse aber als Block von 32 Bytes und um das maximal verwirrend zu machen mit dem obersten Bit der 24 Bit Adresse gesetzt, um dem Kommando zu zeigen, dass das im Flash ist.
Und so wird aus 0x801000 eben 0x800000 + 128 oder auch 0x800000 | (4096 / 32).
Wobei ich die Darstellung in Dezimal einfacher zu benutzen finde da die von EAB generierte .map Datei die Start-Adresse im Flash und die Länge dezimal angibt:
bt817.blob                       : 0      : 4096 
Ubuntu-Light_19_ASTC.glyph       : 4096   : 35904
Ubuntu-Light_19_ASTC.xfont       : 40000  : 309  
Ubuntu-Light_19_ASTC.xfont.pad   : 40309  : 11   
Ubuntu-Light_24_ASTC.glyph       : 40320  : 35904
Ubuntu-Light_24_ASTC.xfont       : 76224  : 309  
Ubuntu-Light_24_ASTC.xfont.pad   : 76533  : 11

**Reply by Marvin1DE** [reply]
ok verstanden danke für deine hilfe,
eine sache aber noch ich dachte die adresse bzw "Ubuntu-Light_19_ASTC.xfont : 40000 : 309" -> 40000 muss zum flashen, ein vielfaches von 4096 sein oder bezieht sich das nur auf den bereich der von Ram_g auf den flash geschoben wird und nicht um die tatsächliche datei.
heißt ich kann mehrere datein in den Ram_g nebeneinander schrittweiße laden und dann vom ram_g in flash nur in im vielfachen von 4096 auf den flash schreiben?

**Reply by RudolphRiedel** [reply]
Beim Lesen und Schreiben gibt es einiges zu beachten, richtig, z.B. schreibt CMD_FLASHUPDATE in vielfachen von 4096.
Die .map Datei da oben kommt aber aus EAB, damit lasse ich mir ein Flash-Image aus mehreren Dateien erzeugen das ich dann entweder direkt per USB/SPI Adapter ins Flash schreibe, oder einmal von meinem Controller schreiben lasse.
Wenn ich über den Controller gehe, dann komprimiere ich das Image vorher noch um es mit CMD_INFLATE zu benutzen.

**Reply by Marvin1DE** [reply]
habs hinbekommen danke für deine erklärungen, sd -> mcu(array) -> ram_g -> flash, funktioniert zumindest jetzt mit mehrere datein und auch datein die größer sind als 4096.
